packages/react-combobox/src/utils/useOptionCollection.ts

```diff
@@ -1,91 +1,71 @@
-        return React.cloneElement(child, overrideProps);
-      }
-
-      // if the child is an option group, get its children
-      // comboboxes semantically only support one level of option groups
-      else if (fluentComponentType === 'OptionGroup' && child.props.children) {
-        const { keys: groupKeys, children: groupChildren } = getValidOptions(child.props.children);
-        keys.push(...groupKeys);
-        return React.cloneElement(child, {}, groupChildren);
-      } else {
-        return React.cloneElement(child);
-      }
-    }
-  });
-
-  return { keys, children: clonedChildren };
-}
+import type { OptionCollectionState, OptionValue } from './OptionCollection.types';
 
 /**
  * A hook for managing a collection of child Options
  */
-export const useOptionCollection = (children: React.ReactNode): OptionCollectionState => {
-  const optionData: React.MutableRefObject<OptionData> = React.useRef({});
-
-  const { collectionData, processedChildren } = React.useMemo(() => {
-    const { keys: optionKeys, children: clonedChildren } = getValidOptions(children);
+export const useOptionCollection = (): OptionCollectionState => {
+  const nodes = React.useRef<{ option: OptionValue; element: HTMLElement }[]>([]);
 
-    const getOptionAtIndex = (index: number) => {
-      const key = optionKeys[index];
-      return optionData.current[key];
-    };
-    const getIndexOfKey = (id: string) => optionKeys.indexOf(id);
-
-    const getOptionByKey = (key: string) => {
-      return optionData.current[key];
+  const collectionData = React.useMemo(() => {
+    const count = () => nodes.current.length;
+    const getOptionAtIndex = (index: number) => nodes.current[index]?.option;
+    const getIndexOfId = (id: string) => nodes.current.findIndex(node => node.option.id === id);
+    const getOptionById = (id: string) => {
+      const item = nodes.current.find(node => node.option.id === id);
+      return item?.option;
     };
 
     return {
-      collectionData: {
-        count: optionKeys.length,
-        getOptionAtIndex,
-        getIndexOfKey,
-        getOptionByKey,
-      },
-      processedChildren: clonedChildren,
+      count,
+      getOptionAtIndex,
+      getIndexOfId,
+      getOptionById,
     };
-  }, [children]);
+  }, []);
 
   const registerOption = React.useMemo(() => {
-    const register = (option: OptionValue) => {
-      if (option && option.key) {
-        optionData.current[option.key] = option;
+    const register = (option: OptionValue, element: HTMLElement) => {
+      const index = nodes.current.findIndex(node => {
+        if (!node.element || !element) {
+          return false;
+        }
+
+        if (node.option.id === option.id) {
+          return true;
+        }
+
+        // use the DOM method compareDocumentPosition to order the current node against registered nodes
+        return Boolean(node.element.compareDocumentPosition(element) === Node.DOCUMENT_POSITION_PRECEDING);
```

Comment:
I find myself a bit wary of this because we're maintaining internal component state based on the state of rendered DOM which feels backward from the typical React equation of `UI = f(state)` (where `UI` is the rendered DOM).

A similar pattern has proven to be a problem Fluent UI v8's `List` virtualization implementation where DOM measurements are cached in React state but not always updated, leading to scrolling issues.

That said, I don't see any specific issues here, plus this is an internal implementation detail so it could be changed later should it cause problems.

Reply:
IMO most of the DOM related stuff on our end tends to be because of our interesting perf stance which is to reduce as many rerenders as possible. React would just say that re-renders are cheap and you should be happy to use them. It's a similar situation with overflow. 

However I do agree, that we should always be skeptical when we see native DOM being used in react